ktor-client/ktor-client-cio/common/src/io/ktor/client/engine/cio/Endpoint.kt

```diff
@@ -30,45 +31,52 @@ internal class Endpoint(
 ) : CoroutineScope, Closeable {
     private val address = NetworkAddress(host, port)
 
+    private val mutex = Mutex()
     private val connections: AtomicInt = atomic(0)
-    private val tasks: Channel<RequestTask> = Channel(Channel.UNLIMITED)
     private val deliveryPoint: Channel<RequestTask> = Channel()
     private val maxEndpointIdleTime: Long = 2 * config.endpoint.connectTimeout
 
-    private val postman = launch {
+    val lastActivity = atomic(GMTDate())
+
+    private val timeout = launch(coroutineContext + CoroutineName("Endpoint timeout($host:$port")) {
         try {
             while (true) {
-                val task = withTimeout(maxEndpointIdleTime) {
-                    tasks.receive()
-                }
-
-                try {
-                    if (!config.pipelining || task.request.requiresDedicatedConnection()) {
-                        makeDedicatedRequest(task)
-                    } else {
-                        makePipelineRequest(task)
-                    }
-                } catch (cause: Throwable) {
-                    task.response.completeExceptionally(cause)
-                    throw cause
-                }
+                delay(maxEndpointIdleTime)
+                if (lastActivity.value + maxEndpointIdleTime < GMTDate()) break
             }
         } catch (cause: Throwable) {
         } finally {
             deliveryPoint.close()
-            tasks.close()
             onDone()
         }
     }
 
-    public suspend fun execute(
+    private suspend fun processTask(task: RequestTask) {
+        try {
+            if (!config.pipelining || task.request.requiresDedicatedConnection()) {
+                makeDedicatedRequest(task)
+            } else {
+                makePipelineRequest(task)
+            }
+        } catch (cause: Throwable) {
+            task.response.completeExceptionally(cause)
+            throw cause
+        }
+    }
+
+    suspend fun execute(
         request: HttpRequestData,
         callContext: CoroutineContext
     ): HttpResponseData {
-        val result = CompletableDeferred<HttpResponseData>()
-        val task = RequestTask(request, result, callContext)
-        tasks.offer(task)
-        return result.await()
+        val response = mutex.withLock {
+            lastActivity.value = GMTDate()
```

Comment:
Should we do that inside of the block or outside (before entering `withLock`) ?

Reply:
We can drop mutex here

Reply:
I replaced `channel` with mutex cause it was `id` change, but since the separate timeout coroutine, we can drop it